CoreFoundation/cmake/modules/CoreFoundationAddFramework.cmake

```diff
@@ -0,0 +1,63 @@
+
+include(CMakeParseArguments)
+
+function(add_framework NAME)
+  set(options STATIC SHARED)
+  set(single_value_args MODULE_MAP)
+  set(multiple_value_args PRIVATE_HEADERS PUBLIC_HEADERS SOURCES)
+  cmake_parse_arguments(AF "${options}" "${single_value_args}" "${multiple_value_args}" ${ARGN})
+
+  set(AF_TYPE)
+  if(AF_STATIC)
+    set(AF_TYPE STATIC)
+  elseif(AF_SHARED)
+    set(AF_TYPE SHARED)
+  endif()
+
+  if(AF_MODULE_MAP)
+    file(COPY
+           ${AF_MODULE_MAP}
+         DESTINATION
+           ${CMAKE_BINARY_DIR}/${NAME}.framework/Headers/${NAME}
+         NO_SOURCE_PERMISSIONS)
+  endif()
+  if(AF_PUBLIC_HEADERS)
+    file(COPY
```

Comment:
do the header copy phases get run before the compile sources phase? that is a strict requirement for clean builds and non darwin builds

Reply:
cc @parkera do you have any thoughts on this Tony?

Reply:
also cc @millenomi what are your thoughts Lily? do you think this is more maintainable?

Reply:
Yeah, the headers get copied before the build into a framework layout of:

```
    CoreFoundation.framework/Headers/...
    CoreFoundation.framework/PrivateHeaders/...
```

This can be changed if needed.

Reply:
the important point here is that before any C file is compiled it needs to copy the headers into place so they can be properly included. and that copy phase could be just a temporary copy phase into a placeholder directory if needed. e.g. `$build_dir/usr/include/CoreFoundation` for the public headers and `$build_dir/usr/local/include/CoreFoundation` for private headers to fake out the `#include <CoreFoundation/CoreFoundation.h>` style includes.

Reply:
Oh, interesting, you would prefer the traditional Unix style layout?  Currently, I actually create a Framework style layout to permit the `<CoreFoundation/CoreFoundation.h>` style inclusion.

Reply:
-F is a Darwin only thing iirc

Reply:
Sure, but, that should be easy enough to extend to other targets.  Its just a header search order thing.  It would mean that this can be built similarly on all the targets.  I can probably do that change in clang pretty easily.

Reply:
It could be but that might be a tall order to get accepted I would guess. It also impacts linking too so you would need to fix gold or switch over to lld. In my opinion the path of least resistance is to just make a library and not a framework.